galley/pkg/config/analysis/diag/message.go

```diff
@@ -59,6 +60,8 @@ type Message struct {
 
 	// DocRef is an optional reference tracker for the documentation URL
 	DocRef string
+
+	Line int
```

Comment:
I think this doesn't work because Origin structure is in the resource.Instance. Different messages may share the same resource.Instance, and this is why the starting line number always works because different messages will show the same line number. However, if we change the line number in the Reference structure to be the exact line, whenever the new message shares the same resource with the former message, the line number in the Reference structure will be replaced to the newer message's error line number.

Reply:
Good point. Here's what we can do instead: the Reference interface is only used in analysis, so let's extend it. Right now it only has a String() method. The high level problem we're running into here is that the error message and the line number aren't encapsulated where we need them--the only way we can get the line number out of the Position struct is via String, because that's all the interface has. I propose:

```golang
type Reference interface {
  String() string
  References() map[string]Reference
}
```
Now when we're iterating through the yaml and recursing through all of the Nodes, we can create a Position struct for each level and create a matching structure with Positions. i.e. this:
```yaml
parent:
  levelOne:
    - levelTwo: a
    - levelTwo: b
```
would become something like
```
Position {filename, 0, map[string]Position {
  'levelOne': Position{filename, 1, map[string]Position {
    '0': Position{filename, 2, map[string]Position{}},
    '1': Position{filename, 3, map[string]Position{}},
  },
}
```
Then in the individual ErrorLine functions, instead of building the entire string like
`"parent.levelOne[%d]"`, we can first look for "parent", then recursively look for "levelOne", and so on until we find the key we're looking for (or reach an error). I admit that the handling of arrays as well as maps is a bit inelegant, but we can start from here and make it look nicer later.

This should require minimal refactoring but also allow us to keep the same printing logic (since we're still printing a Reference at the end), and since the line number is part of the struct we don't need to search a string to replace it.

Reply:
Looks like it's better than writing those utility functions. Thanks a lot for providing this idea! However, can you clarify how to use this nested Reference? From my understanding, the analyzers' developers are going to provide the path variables for the message, and during the message rendering process, the path variables are going to be used to search for the correct Reference.

For example, if a pathKey is originally like "parent.childOne[0].childTwo", if we want to obtain the Reference of childTwo, are we going to pass []interface {"parent", "childOne", 0, "childTwo"} in the analyzer as the variable for the error message, and later use a search function with these variables to find the right Reference to use the String() method?
Is this the correct implementation or these are other ways to finally find the exact Reference?

Reply:
Yep, that's the basic idea. We could pass the pathKey just like that and split it later--I think you already wrote the logic for parsing the path strings, and I think it would be easier for developers to put a string in instead of a list of strings. i.e. maybe something like

```go
const mirrorPath = "{.spec.http[%d].mirror.host}"
...
func (a *DestinationHostAnalyzer) analyzeVirtualService...
...
if line, ok := util.ErrorLine(r, fmt.Sprintf(mirrorPath, d.ServiceIndex)); ok {
				m.Line = line
			}
```


I don't think we should use interface{} though--just make everything a string. Technically someone could write something like
```yaml
key[3]: value
```
which might not work, but let's not worry about that for the first phase. If it ever becomes an issue we can just use escape codes or something.

Reply:
This idea is currently not applied because it only helps the analyzer to display file names and line numbers more easily. However, implementing this idea will make the code structure of the reference part complicated. At the same time, in order to use this structure, we need to parse the complete path into single fields, and then refer to the keys in the Reference map to find the line number. The current method utilizes the string template: the developer only needs to provide the necessary parameters, and when the complete path is available, you can directly search the corresponding line number in the flat field map, which is easier and simpler. However, the idea mentioned above can be pretty useful if there are some simper implementations or different ways to utilize it. It may be more extensible compared to adding string templates.